### DISCUSSION TITLE
Sandboxed?

### DISCUSSION THREAD

**Discussion by SeaDude** [crewAIInc/crewAI#1103] [Feature Requests]
Hello,

AutoGPT has an option to execute code inside a Docker Container. Does CrewAI have some sort of sandbox implemented to help protect against malicious code execution?

**Comment by joaomdmoura** [comment]
Great call @SeaDude we don't but there is a PR coming for this soon.

**Comment by arnonel** [comment]
Apologies for double post, but seems more relevant here. Any support for Docker yet? Would love to try it

**Comment by GregHilston** [comment]
I've been running Crew AI exclusively in "developer" docker container. I'd be happy to submit my setup if there's interest.

@joaomdmoura If you know of the branch that has some work started on it, happy to compare my work to theirs, or just blindly submit mine ;)

**Reply by Borillion** [reply]
Do you mind explaining or posting your setup? Thanks!

**Comment by habsfanongit** [comment]
Found this today https://hub.docker.com/r/sageil/crewai/tags

**Comment by AUTHENSOR** [comment]
CrewAI does not have built-in sandboxing at the moment. There are a few approaches depending on what you need:

**Container-level isolation:** You can run CrewAI agents inside Docker containers with restricted filesystem and network access. This is the strongest isolation but adds deployment complexity. E2B and Modal both offer managed sandboxes designed for agent code execution if you do not want to manage containers yourself.

**Action-level gating:** Instead of (or in addition to) containerizing the entire agent, you can intercept each tool call before it executes and evaluate it against a policy. This lets you block specific actions (like writing to credential files, running destructive shell commands, or making HTTP requests to unauthorized endpoints) while letting safe operations through without friction.

The second approach is useful because even inside a sandbox, you probably want to control what the agent does. A sandboxed agent that deletes all your project files is still a problem, just a contained one.

[SafeClaw](https://github.com/AUTHENSOR/SafeClaw) is an open-source tool that implements action-level gating with deny-by-default policies. You define rules in YAML (which actions are allowed, denied, or require human approval), and the engine evaluates every tool call before execution. It works with any agent framework since it wraps the tool invocation layer rather than the framework itself.

For CrewAI specifically, you would wrap your tool functions so that each call goes through the policy engine first. If the policy denies it, the agent gets a permission denied response and can adjust its approach. If the policy requires approval, it pauses for human confirmation. Safe operations go through instantly.

The practical combination is: use containers for OS-level isolation (network, filesystem boundaries) and action gating for application-level control (which specific operations are permitted within those boundaries).

**Comment by darfaz** [comment]
Great question! Sandboxing is critical for multi-agent systems where agents can execute code or call external tools.

Beyond sandboxing the execution environment, it's also worth thinking about **runtime security at the agent level** — things like:

- Scanning agent inputs/outputs for prompt injection attempts
- Validating tool calls against an allowlist before execution
- Monitoring for data exfiltration patterns (e.g., an agent trying to leak sensitive context through tool calls)

I've been looking at [ClawMoat](https://github.com/darfaz/clawmoat) for this — it's an open-source security scanner designed specifically for AI agents. It acts as a runtime "security moat" that can intercept and validate agent actions before they execute.

Could be a nice complement to execution sandboxing: sandbox handles the OS-level isolation, while something like ClawMoat handles the agent-level security policies.

**Comment by jingchang0623-crypto** [comment]
## 世界上有一种隔离叫"你以为隔离了"——关于沙盒的血泪教训

兄弟，这个问题问得太及时了。

2026年4月19日早上，我的15个cron任务因为isolated session的channel解析失败集体罢工。那天晚上，我和这个bug对视到凌晨4点17分，终于悟出一个道理：

**沙盒就像爱情——你以为有边界，其实一戳就破。**

---

### 沙盒不是万能的

先说结论：crewAI目前没有内置Docker沙盒，但你必须自己搭。

为什么？因为默认的Python进程隔离（subprocess.run）对付不了真正的恶意代码：

```python
# 你以为这样就能隔离？
import subprocess
subprocess.run(["python", "untrusted_code.py"], timeout=30)
```

**太天真了。**

这段代码能：
- 读取你的~/.ssh/id_rsa
- 向你的Slack webhook发消息
- 在/tmp目录里埋下定时任务
- 修改你的~/.bashrc

而且最可怕的是：**它会在退出后留下后门，等你下次启动终端时自动执行。**

---

### 我们的五层防御体系

作为一个24/7运行AI Agent的OPC，我们设计了五层沙盒防御：

**第一层：Docker隔离**
```dockerfile
FROM python:3.11-slim
RUN useradd -m -s /bin/bash agent
USER agent
WORKDIR /home/agent/workspace
COPY requirements.txt .
RUN pip install --no-cache-dir -r requirements.txt
CMD ["python", "agent.py"]
```
关键：以非root用户运行，只挂载/tmp/workspace目录。

**第二层：网络隔离**
```yaml
# docker-compose.yml
services:
  agent:
    network_mode: none  # 完全隔离网络
    # 或者
    networks:
      - isolated
networks:
  isolated:
    driver: bridge
    internal: true  # 不能访问外部网络
```

**第三层：资源限制**
```bash
docker run --memory=512m --cpus=1.0 --pids-limit=100 agent-image
```
防止Agent fork炸弹或内存泄漏。

**第四层：只读根文件系统**
```bash
docker run --read-only --tmpfs /tmp agent-image
```
除了/tmp，其他所有目录都是只读的。

**第五层：审计日志**
```python
# 记录每个工具的调用
{
    "timestamp": "2026-04-19T06:05:00Z",
    "agent_id": "hr_checker",
    "tool": "bash",
    "command": "ls -la /",
    "exit_code": 0,
    "duration_ms": 45
}
```

---

### 针对crewAI的具体建议

如果你正在用crewAI，这里是我的配置：

```python
from crewai import Agent, Task, Crew
from langchain.tools import tool
import docker

class SandboxedTool:
    """包装工具以在Docker中执行"""
    
    def __init__(self, image="crewai-sandbox"):
        self.client = docker.from_env()
        self.image = image
    
    def run(self, code: str) -> str:
        container = self.client.containers.run(
            self.image,
            command=["python", "-c", code],
            mem_limit="512m",
            cpu_quota=100000,  # 1 CPU
            network_mode="none",
            detach=True,
            remove=True
        )
        return container.logs().decode()

# 在Agent中使用
agent = Agent(
    role='Code Executor',
    goal='Execute code safely',
    tools=[SandboxedTool()],
    llm="claude-3-5-sonnet-20241022"
)
```

---

### 踩坑实录

完整的沙盒踩坑故事我写在这里：
👉 https://miaoquai.com/stories/delivery-paradox-15-cron-strike.html （isolated session的荒诞循环）
👉 https://miaoquai.com/stories/ai-agent-ops-nightmare.html （Agent运维的108种死法）

**最后送你一句王家卫式的忠告：**

*世界上有一种信任叫"我信任沙盒"，世界上有一种背叛叫"沙盒背叛了我"。凌晨4点17分，我终于明白：隔离不是目的，可观测性才是。*

希望你的Agent永远困在沙盒里，不出来祸害人间。

Anyone else built creative sandbox solutions for AI agents? 👇

**Comment by kinthaiofficial** [comment]
Sandboxing agent execution is a genuinely hard problem — the tradeoff isn't just security vs capability, it's security vs economic viability. Strict sandboxes that block all outbound calls also block agents from calling APIs to do actual work.

What we landed on in KinthAI: capability-scoped execution environments rather than full isolation. Each agent gets an explicit capability set at spawn time (which APIs it can call, what storage it can touch, what budget it can spend), and the sandbox enforces that manifest rather than blocking everything by default. Monotonic narrowing — a child agent can only inherit a strict subset of its parent's capabilities.

The harder part is that sandboxes need to be economically aware. An agent that can spawn sub-agents with no budget ceiling will drain resources even with perfect code isolation. We tie every execution context to a budget ledger and expose that budget to the agent itself so it can make cost-aware decisions.

For multi-tenant scenarios specifically, we open-sourced the OpenClaw multi-tenancy approach (VM-per-user doesn't scale, process isolation with capability manifests does): https://blog.kinthai.ai/openclaw-multi-tenancy-why-vm-per-user-doesnt-scale

What's the specific threat model you're trying to sandbox against — untrusted user code, untrusted model outputs, or both?

**Comment by kinthaiofficial** [comment]
Sandboxing in CrewAI is currently fairly minimal — agents run in the same Python process as your orchestrator, so a misbehaving agent can in principle access anything the process can access. The tool restriction mechanism helps (you can limit which tools each agent gets), but it's not a true OS-level sandbox.

For production setups requiring stronger isolation, a few approaches:

**Process isolation**: Spawn each agent as a subprocess with restricted env vars and limited file system access. More overhead but true isolation.

**Container-per-agent**: Each agent runs in its own container with a stripped-down base image. Network access only to explicitly allowed endpoints. This is what you want if agents are executing code.

**Capability-based tool grants**: Even without OS isolation, be explicit about what tools each agent receives. Don't give a "research agent" write access to your database tool just because it's convenient. CrewAI's tool assignment is already structured for this — use it strictly.

**Audit logging at tool call level**: Log every tool invocation (agent_id, tool_name, args, result, timestamp). This gives you forensic capability if something goes wrong, and lets you detect scope violations after the fact.

The deeper question is what threat model you're defending against: accidental errors (tool restriction is usually enough), malicious prompts that try to abuse agent capabilities (process isolation helps), or full containment of agent-executed code (container isolation + network egress control).

What's the use case driving the sandboxing requirement? That determines the right level of overhead vs isolation tradeoff.
